Make evil-cp themed key bindings more practical

Somelauw opened the pull request:
- Duplicate key bindings have been removed.
- Key bindings y, < and > are not bound unconditionally.
- New evilcp-special key theme for barfing and slurping like in evil-cleverparens.

abo-abo commented:
Thanks. Seems you rebind a lot of functions in one commit. Could you make a summary of the rebinds, explaining how each one is an improvement?

I don't use `evil-cp` theme, but I can merge this commit if I get an OK from @ryepesg or other `evil-cp` theme users.

Somelauw commented on lispy.el:
```diff
@@ -8563,14 +8563,19 @@ When ARG is non-nil, unquote the current string."
 
 (defvar lispy-mode-map-evilcp
   (let ((map (copy-keymap lispy-mode-map-base)))
-    (define-key map (kbd "M-)") 'lispy-close-round-and-newline)
```
evil-cleverparens defines M-) to wrap the previous s-expression

ryepesg replied:
I agree, it could be bound to `evil-cp-wrap-previous-round` or equivalent in lispy.

Somelauw commented on lispy.el:
```diff
@@ -8563,14 +8563,19 @@ When ARG is non-nil, unquote the current string."
     (define-key map (kbd "}") 'lispy-close-curly)
     (define-key map (kbd ")") 'lispy-right-nostring)
     (define-key map (kbd "\"") 'lispy-doublequote)
+    ;; wrapping
```
wrapping commands are backed by lispy, not by evil-cleverparens or paredit (no need to install these)

Somelauw commented on lispy.el:
```diff
@@ -8579,23 +8584,18 @@ When ARG is non-nil, unquote the current string."
     (define-key map (kbd "M-?") 'lispy-convolute-sexp)
     (define-key map (kbd "M-S") 'lispy-split)
     (define-key map (kbd "M-J") 'lispy-join)
-    (define-key map (kbd "{") 'lispy-braces)
```
These bindings directly conflict with bindings right above. The bindings defined first follow the same rules as paredit. The bindings defined here behave like lispy. I have followed paredit conventions, because cleverparens by default doesn't have special behaviour for square brackets.

I can make a separate evil-cp-lispy or evil-cp-paredit theme if desired or even remove all bindings except the ones with meta, so this can be combined with the themes for paredit, lispy or even parinfer depending on how someone wants [ and ] to behave. On the other hand, evil-cp/paredit users may want M-] and evil-cp/lispy users may want M-} to wrap something in square brackets ("M-]" is the standard binding in evil-cleverparens).

Somelauw commented on lispy.el:
```diff
@@ -8579,23 +8584,18 @@ When ARG is non-nil, unquote the current string."
-    (define-key map (kbd "M-(") 'evil-cp-wrap-next-round)
-    (define-key map (kbd "M-{") 'evil-cp-wrap-next-curly)
-    (define-key map (kbd "M-}") 'evil-cp-wrap-next-square)
-    (define-key map (kbd "<") 'evil-cp-<)
```
Having <, > and y bound in insert mode (not just in special positions) isn't done by cleverparens and is undesirable in my opinion. It also shadows the binding for lispy-occur. I have added a separate evil-cp-special theme to make < and > behave like evil-cleverparens in special positions.

Somelauw commented on lispy.el:
```diff
@@ -8579,23 +8584,18 @@ When ARG is non-nil, unquote the current string."
+(defvar lispy-mode-map-evilcp-special
+  (let ((map (copy-keymap lispy-mode-map-special)))
+    (lispy-define-key map ">" 'lispy-slurp-or-barf-right)
+    (lispy-define-key map "<" 'lispy-slurp-or-barf-left)
```
This is how < and > behave in normal mode with evil-cleverparens enabled.

Somelauw commented on lispy.el:
```diff
@@ -8579,23 +8584,18 @@ When ARG is non-nil, unquote the current string."
-    (define-key map (kbd "M-}") 'evil-cp-wrap-next-square)
-    (define-key map (kbd "<") 'evil-cp-<)
-    (define-key map (kbd ">") 'evil-cp->)
-    (define-key map (kbd "y") 'lispy-new-copy)
```
Having <, > and y bound in insert mode (not just in special positions) isn't done by cleverparens and is undesirable in my opinion. It also shadows the binding for lispy-occur. I have added a separate evil-cp-special theme to make < and > behave like evil-cleverparens in special positions.

ryepesg replied:
I agree it is not convenient. To make it consistent with both paredit and cleverparens, what do you think bout:

```
     (define-key lispy-mode-map (kbd "C-<") 'evil-cp-<)
     (define-key lispy-mode-map (kbd "C->") 'evil-cp->)
```

Somelauw commented on lispy.el:
```diff
@@ -8579,23 +8584,18 @@ When ARG is non-nil, unquote the current string."
-    (define-key map (kbd "{") 'lispy-braces)
-    (define-key map (kbd "}") 'lispy-brackets)
-    (define-key map (kbd "]") 'lispy-forward)
-    (define-key map (kbd "[") 'lispy-backward)
```
These bindings directly conflict with bindings right above. The bindings defined first follow the same rules as paredit. The bindings defined here behave like lispy. I have followed paredit conventions, because cleverparens by default doesn't have special behaviour for square brackets.

I can make a separate evil-cp-lispy or evil-cp-paredit theme if desired or even remove all bindings except the ones with meta, so this can be combined with the themes for paredit, lispy or even parinfer depending on how someone wants [ and ] to behave. On the other hand, evil-cp/paredit users may want M-] and evil-cp/lispy users may want M-} to wrap something in square brackets ("M-]" is the standard binding in evil-cleverparens).

ryepesg commented on lispy.el:
```diff
@@ -8563,14 +8563,19 @@ When ARG is non-nil, unquote the current string."
     (define-key map (kbd "[") 'lispy-open-square)
     (define-key map (kbd "]") 'lispy-close-square)
     (define-key map (kbd "{") 'lispy-open-curly)
     (define-key map (kbd "}") 'lispy-close-curly)
```
I never use `lispy-close-square` or `lispy-close-curly` if I can maintain  everything balanced. What do yo think about:

```
     (define-key lispy-mode-map (kbd "]") 'lispy-braces)
     (define-key lispy-mode-map (kbd "[") 'lispy-brackets)
     (define-key lispy-mode-map (kbd "}") 'lispy-forward)
     (define-key lispy-mode-map (kbd "{") 'lispy-backward)
```

Somelauw replied:
I you are okay with keeping [ and ] for lispy-forward and lispy-backward, I have made two proposals below that let you do this.

ryepesg commented on lispy.el:
```diff
@@ -8579,23 +8584,18 @@ When ARG is non-nil, unquote the current string."
-    (define-key map (kbd "}") 'lispy-brackets)
-    (define-key map (kbd "]") 'lispy-forward)
-    (define-key map (kbd "[") 'lispy-backward)
-    (define-key map (kbd "M-(") 'evil-cp-wrap-next-round)
```
What do you think about:
```
(define-key lispy-mode-map (kbd "M-(") 'lispy-wrap-round)
```

Somelauw replied:
That one is already there, but I moved that line to the top

ryepesg replied:
perfect

Somelauw commented:
I have described the individual changes. One change is that after the pull request keys mostly behave like paredit, whereas originally some keys where double defined. Some people may want the lispy style instead. I have two proposals to solve this.

#### Proposals:
##### Make a separate evil-cp-lispy and evil-cp-paredit theme

evil-cp-lispy follows lispy bindings + meta bindings. evil-cp-paredit follows paredit bindings + meta bindings.

For example:

        (lispy-set-key-theme '(evilcp-special evil-cp-lispy c-digits))
        ;; (lispy-set-key-theme '(evilcp-special evil-cp-paredit c-digits))

##### Make the evil-cp theme stick to M-bindings alone.
Then it can be combined with the themes for paredit, lispy or even parinfer depending on how someone wants [ and ] to behave.

For example

        (lispy-set-key-theme '(evilcp-special lispy evil-cp c-digits))
        ;; (lispy-set-key-theme '(evilcp-special paredit evil-cp c-digits))
        ;; (lispy-set-key-theme '(evilcp-special parinfer evil-cp c-digits))

The disadvantage is that evil-cp/paredit users may want M-] and evil-cp/lispy users may want M-} to wrap something in square brackets. I propose "M-]" as a default, because that's what evil-cleverparens uses.

ryepesg commented on lispy.el:
```diff
@@ -8563,14 +8563,19 @@ When ARG is non-nil, unquote the current string."
+    (define-key map (kbd "M-(") 'lispy-wrap-round)
+    (define-key map (kbd "M-{") 'lispy-wrap-braces)
+    (define-key map (kbd "M-[") 'lispy-wrap-brackets)
     (define-key map (kbd "M-\"") 'lispy-meta-doublequote)
```
`lispy-meta-doublequote` and `paredit-meta-doublequote` are different, the second one wraps the string with double quotes, what I think is better.

Somelauw replied:
You are right. They are different and paredit-meta-doublequote is what we want.

@abo-abo Would it be okay to add something like lispy-wrap-doublequotes (or different function name) to lispy so that your package can be used without paredit?

To illustrate the function (take '|' as the cursor position)

    |this

should be turned into

    "|this"

ryepesg commented:
I like the second proposal (Make the evil-cp theme stick to M-bindings alone) and also I agree with `M-[` to wrap something in square brackets.

In cases of conflict between themes, the order in which the user select those matters for overriding keybindings. It would be different:

(lispy-set-key-theme '(**paredit evil-cp** c-digits))

(lispy-set-key-theme '(**evil-cp paredit** c-digits))

I am not sure if we really need `evilcp-special` or if we can just merge it with `evil-cp` directly.

abo-abo commented:
@Somelauw @ryepesg Please update the PR once you've agreed on the bindings.

Somelauw commented:
I have added what we agree on so far.

> I am not sure if we really need evilcp-special or if we can just merge it with evil-cp directly.

I don't know about this one. Here are some options:
1. Keep evil and evilcp-special apart
1. Merge evil and evilcp-special together (evilcp always give you special)
1. Automatically use evilcp-special when both evilcp and special are specified

I think 1 is most flexible, but maybe 2 or 3 are simpler.

Maybe evilcp-special can also have some other vim-like bindings like 'y' for lispy-new-copy but then probably another key should be bound to lispy-occur etc.

Somelauw commented:
I have changed DEL to lispy-delete-backward instead of lispy-backward-delete, because I think this function is more intuitive when trying to delete empty strings.

Somelauw commented:
Looking at it again, I think lispy-delete-backward-or-splice-or-slurp and lispy-delete-or-splice-or-slurp are the functions closest to how x and X behave in evil-cleverparens.

What I think is most important about this key theme is the behaviour of < and > in special mode. Having meta-bindings is nice but for evil users these are already defined by evil-cleverparens or by lispyville in normal mode. The rest of the stuff might be a matter of preference, so should maybe not be defined at all or be defined to be as similar to evil-cleverparens as possible.

Somelauw commented:
I have added the bracket pairs back to evilcp. This makes it possible to use evilcp on its own. To make it easier to combine evillcp with lispy, I have changed the priority, so that if evilcp and lispy (or paredit) are both specified the latter gets precedence.

Here are two examples:
```
;; Behave as much like cleverparens as possible with special bindings
(lispy-set-key-theme evilcp evilcp-special)
```

```
;; Combine cleverparens with lispy, so that [ and ] can be used to jump to brackets
(lispy-set-key-theme lispy evilcp evilcp-special)
```

@ryepesg I think the second one is the one you want. Let me know if this works.

abo-abo commented:
> Let me know if a new pull request should:

I don't know if `lispy-mode-map-evilcp` is used or relevant anymore. It makes sense to remove it if no one uses it. Please ping the author that introduced it and go ahead with removal if it's fine.

Somelauw commented:
@ryepesg See comment above

ryepesg commented:
We disagree about some configurations, but if I am the only one using evilcp, sure, lets remove it. Also, if you want to use the name evilcp alone (without the "special") feel free to do so.

Please keep making easy for new users to apply their own customisations, so even if a new user has a different view can replace those specific bindings.

abo-abo commented:
@ryepesg If you're still using it, we might as well keep it. Maybe something can be done about the suggestions of @Somelauw that some stuff doesn't work with your theme.

Maybe it's also possible to extract more common stuff from the existing themes to reduce code repetition.

ryepesg commented:
Thank you @abo-abo. For me it is easy to apply my own configurations, so the issue is what is the best for newcomers. Having two similar themes is confusing for them and @Somelauw is a more active contributor, so I think he can continue this tasks forward.
